### DISCUSSION TITLE
Issues with recursive types + metaclasses

### DISCUSSION THREAD

**Discussion by Gobot1234** [microsoft/pyright#2503] [Q&A] [answered]
I have some admittedly rather odd code making use of metaclasses and recursive types for code similar to `(Query.running / 440) & Query.no_password`. I'm not entirely sure if I've typed all of this correctly but I have a few issues and I wanted some feedback on it.

```py
from __future__ import annotations

from typing import Any, Generic, TypeVar
from enum import Enum

from typing_extensions import TypeAlias


T = TypeVar("T")
Q: TypeAlias = "Query[Q]"


class Operator(Enum):
    div = "/"
    and_ = "&"


def process(op, q1, q2) -> Q:
    ...


class QueryMeta(type):
    @property
    def no_password(cls) -> Q:
        return Query[Q]("no_password")

    @property
    def running(cls) -> Query[int]:
        return Query[int]("running")


class Query(Generic[T], metaclass=QueryMeta):
    def __init__(self, *parts: Query[Any] | Operator | str):
        self.parts = parts

    def __divmod__(self, other: T) -> Q:
        return process(Operator.div, self, other)

    def __and__(self, other: Q) -> Q:
        return process(Operator.and_, self, other)
```
Issue 1 - the revealed types appear incorrect
`reveal_type(Query.running)  # revealed type is Any`
`reveal_type(Query.no_password)  # revealed type is Any`

If I add: `running: Query[int] = Query.running`

Issue 2 - type errors are not reported
`running / "not an int"` is Any rather than a type error

Issue 3 - the revealed types for `Q` and `running / 440` are `Type[Query[Q]]` and `Any` respectively neither of which seem correct
Screenshot
![image](https://user-images.githubusercontent.com/50501825/139159319-bd96dfe2-bca5-475c-94e5-0b7728ef9e54.png)

Could you maybe shed some light on why some of this is the way that it is?
Thank you.

**Comment by erictraut** [comment, marked as answer]
Are you using pyright or pylance? The reason I ask is that it appears you have typeCheckingMode set to "off". This is the default for pylance, but pyright defaults to "basic". You should see more errors if type checking is enabled.

There is a bug here. Pyright isn't properly handling properties defined on a metaclass. I'd appreciate it if you'd file a separate bug report for that.

**Reply by Gobot1234** [reply]
Oh I didn't realise pylance didn't enable pyright's type checking by default. Turning it on reveals another issue with the TypeAlias as a return type. 
![image](https://user-images.githubusercontent.com/50501825/139227635-817726cf-2aec-495b-bf83-8ace9b3b7127.png)
I'm also not sure about `Q` showing up as `Type[Query[Q]]` should it not be `Query[Q]`?

I'll make an issue about metaclass properties
